### PR TITLE
refactor(ci+scripts): modular ESM migration, secrets modes, validation & docs overhaul

### PR DESCRIPTION
**Pull request by Xain3**
## Summary

Comprehensive modernization of the project:
- Migrates codebase to ES modules with path alias imports.
- Expands test suite (unit, integration, end-to-end) and tightens coverage enforcement.
- Introduces hardened secrets retrieval (file, external, gcp, azure, aws, auto) and timeout controls.
- Adds schema sync automation, package.json validation workflow, and improved coverage script logic.
- Overhauls contributor + automation documentation and PR template for clarity and consistency.

## Related issue

- Fixes / relates to: (add issue numbers if applicable)

## Implementation notes

- Converted CommonJS to ESM across scripts, helpers, tests.
- Centralized validation logic (`helpers/config-validator.js`) and expanded schema-based + semantic validation tests.
- Added experimental cloud secrets modes with strict argument validation and safe temp file handling.
- Improved coverage discovery (`check-coverage.cjs`) with path search + graceful fallbacks.
- Added monthly schema sync workflow + script; package.json validation workflow.
- Strengthened test matrix: secrets modes, timeout precedence, config caching, CLI flows.
- Documentation rewritten: root README, `.github/` docs, contributing guide, agent instructions, new PR template.

### ESM & Architecture Refactor

- All scripts/tests moved to native ESM imports.
- Introduced path alias usage throughout (`#helpers/*`, `#scripts/*`, etc.).
- Added extensive unit coverage for compose generation, validation, secrets logic.

### Secrets & Compose Enhancements

- Added `gcp|azure|aws|external|file|none|auto` secrets modes.
- Timeout override precedence: CLI flag > env > default (8000ms).
- Secure temp file handling + explicit error wrapping for cloud CLIs.

### Validation & Schema Tooling

- Expanded `validate-config` with cache integrity handling, schema error surfacing, and richer CLI UX.
- Added `validate-package-json.js` + shell wrapper + CI workflow.
- Added schema sync script + scheduled workflow (SchemaStore mirror).

### CI / Automation

- Coverage script now dynamically locates `coverage-summary.json` with bounded search depth and env override.
- Added coverage threshold entry for `helpers/`.
- Jest editor settings + lightweight ESLint config committed.

### Documentation & Templates

- New `CONTRIBUTING.md` (stepwise journey + reference).
- Updated root `README.md` with concepts, CLI reference, secrets modes, path aliases, coverage, builder notes.
- Replaced PR template with structured checklist + clearer review guidance.
- Added `.github/README.md` describing automation assets.
- Refreshed AI agent instructions for repository-specific conventions.

### Testing Expansion

- Added granular tests: secrets mode resolution, timeout precedence, cloud secret retrieval stubs, config caching corruption scenarios, CLI argument parsing edge cases.
- Added raw YAML stub + shell script integration tests.

## Testing

- All Jest unit, integration, and end-to-end tests pass locally and in CI.
- New tests cover previously uncovered branches (secrets modes, cache invalidation, schema mismatch).
- Coverage thresholds maintained; helpers directory threshold added.

## How to verify locally

```zsh
npm install
npm run test:ci
npm run check-coverage
# Generate compose (sample)
npx fvtt-compose-gen -c container-config.example.json -o compose.dev.yml
npx fvtt-pod up -d

### CODE REVIEW

**Comment by Xain3** [comment]
@copilot increase test coverage for scripts, as currently it fails.

**Comment by Xain3** [comment]
@copilot increase test coverage for scripts, as currently it fails.

**Comment by Xain3** [comment]
@copilot  increase test coverage for scripts, as currently it fails.

File: tests/unit/scripts/validate-config.unit.test.js [hunk 1]
```diff
@@ -0,0 +1,918 @@
...
+  checkConfigWithCache,
+  parseCommandLineArgs,
+  showHelpMessage
+} from '../../../scripts/validate-config.js';
+
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
+
+function runNode(args, opts = {}) {
+  try {
+    return childProcess.execSync(`node ${args}`, { encoding: 'utf8', stdio: ['ignore', 'pipe', 'pipe'], ...opts });
```
[the remark this answers is not in the document]
Comment (Xain3) [reply tests/unit/scripts/validate-config.unit.test.js]: Outdated

File: scripts/generate-compose.js [hunk 2]
```diff
@@ -144,12 +198,84 @@
...
+	return execSync(gcpCommand, { encoding: 'utf8' });
+}
+
+function retrieveAzureSecret(vaultName, secretName) {
+	const azureCommand = `az keyvault secret show --vault-name "${vaultName}" --name "${secretName}" --query value --output tsv`;
+	return execSync(azureCommand, { encoding: 'utf8' });
+}
+
+function retrieveAwsSecret(region, secretName) {
+	const awsCommand = `aws secretsmanager get-secret-value --region "${region}" --secret-id "${secretName}" --query SecretString --output text`;
```
[the remark this answers is not in the document]
Comment (Xain3) [reply scripts/generate-compose.js]: outdated

File: scripts/generate-compose.js [hunk 3]
```diff
@@ -144,12 +198,84 @@
...
 	};
 }
 
+function retrieveGcpSecret(project, secretName) {
+	const gcpCommand = `gcloud secrets versions access latest --secret="${secretName}" --project="${project}"`;
+	return execSync(gcpCommand, { encoding: 'utf8' });
+}
+
+function retrieveAzureSecret(vaultName, secretName) {
+	const azureCommand = `az keyvault secret show --vault-name "${vaultName}" --name "${secretName}" --query value --output tsv`;
```
[the remark this answers is not in the document]
Comment (Xain3) [reply scripts/generate-compose.js]: Outdated

File: tests/integration/fvtt-compose-gen.int.test.js [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+ * @param {object} opts - Additional options for execSync
+ * @returns {object} - Object with stderr and error info
+ */
+function runComposeGenExpectFailure(args = [], opts = {}) {
+  const repoRoot = path.resolve(__dirname, '../..');
+  const binaryPath = path.join(repoRoot, 'scripts/generate-compose.js');
+  const cmd = `node ${binaryPath} ${args.join(' ')}`;
+  
+  try {
+    childProcess.execSync(cmd, { encoding: 'utf8', stdio: 'pipe', ...opts });
```
[the remark this answers is not in the document]
Comment (Xain3) [reply tests/integration/fvtt-compose-gen.int.test.js:41]: Used in tests, not in production.

File: tests/integration/end-to-end-workflow.int.test.js [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+ * Run the fvtt-compose-gen binary with arguments
+ * @param {string[]} args - Command line arguments
+ * @param {object} opts - Additional options for execSync
+ * @returns {string} - Standard output from the command
+ */
+function runComposeGen(args = [], opts = {}) {
+  const repoRoot = path.resolve(__dirname, '../..');
+  const binaryPath = path.join(repoRoot, 'scripts/generate-compose.js');
+  const cmd = `node ${binaryPath} ${args.join(' ')}`;
+  return childProcess.execSync(cmd, { encoding: 'utf8', ...opts });
```
[the remark this answers is not in the document]
Comment (Xain3) [reply tests/integration/end-to-end-workflow.int.test.js:27]: Used in tests

File: scripts/generate-compose.js [hunk 6]
```diff
@@ -92,269 +117,542 @@
...
+	toEnvList,
+	resolveTemplatedString,
+	resolveTemplatedNumber,
+	buildComposeFromComposeConfig,
+	buildComposeFromContainerConfig,
+	main
+};
+
+if (import.meta.url === `file://${process.argv[1]}`) {
+  try { main(); } catch (e) { console.error(e?.stack || String(e)); process.exit(1); }
```
[the remark this answers is not in the document]
Comment (Xain3) [reply scripts/generate-compose.js:736]: SHould be solved

**Comment by Xain3** [comment]
@copilot CI: unify coverage run and make checker robust 

 --- 

 Summary

- Fix CI coverage failures by generating a single consolidated Jest coverage report and making the coverage checker robust to CI layout differences.

What changed

- `.github/scripts/check-coverage.cjs`
  - Now searches multiple candidate paths for `coverage/coverage-summary.json`.
  - Respects `COVERAGE_SUMMARY` env var if provided.
  - Tries repo-root-derived paths and performs a bounded repo scan (skips `node_modules` and `.git`) as a final fallback.
  - Prints attempted paths when it fails to help CI diagnostics.

- `package.json`
  - Added `test:combined` (runs jest across `tests/unit` and `tests/integration` with `--coverage`).
  - `test:ci` now delegates to `test:combined`.

- Docs
  - `docs/ci-and-coverage.md` and `README.md` updated to document `test:combined`/`test:ci` usage and the updated check-coverage behavior.

Why

- Previously coverage was only produced by one of the two separate Jest runs which caused `check-coverage` to sometimes miss `coverage-summary.json` in different CI layouts. Running Jest once for both test sets ensures a single coverage summary. The updated checker improves resilience and diagnostics.

Testing performed

- Ran `npm run test:ci` locally (unified jest run) — all test suites passed and coverage report was generated.
- Ran `node .github/scripts/check-coverage.cjs` locally — checker located the summary and reported per-folder thresholds met.

How to reproduce locally

```zsh
npm install
npm run test:ci
node .github/scripts/check-coverage.cjs
```

Notes / follow-ups

- If you want I can also update the GitHub Actions workflow to call `npm run test:combined` explicitly, but `test:ci` already delegates to it.
- I avoided changing the developer-facing `test:unit` and `test:integration` scripts so local iteration remains fast.

Signed-off-by: automated copilot agent

File: tests/unit/patches/entrypoint/wrappers.dryrun.unit.test.js [hunk 7]
```diff
@@ -71,10 +70,8 @@ describe('wrapper scripts dry-run', () => {
     });
 
     test('target .mjs file exists', () => {
-      // verify the target .mjs file exists
-  const fs = require('node:fs');
-  const exists = fs.existsSync(expectedViaWrapperAbs);
-  expect(exists).toBe(true);
+      const exists = fs.existsSync(expectedViaWrapperAbs);
+      expect(exists).toBe(true);
```
[the remark this answers is not in the document]
Comment (Xain3) [reply tests/unit/patches/entrypoint/wrappers.dryrun.unit.test.js]: Implemented in 368682f .

File: scripts/generate-compose.js [hunk 8]
```diff
@@ -92,269 +117,652 @@
...
+    return {
+      topLevel: { [name]: { external: true } },
+      serviceRef: [ { source: name, target: opts.secretsTarget || 'config.json' } ],
+    };
+  }
+
+  if (mode === 'gcp' || (mode === 'auto' && opts.secretsGcpProject && opts.secretsGcpSecret)) {
+    console.warn('[experimental] GCP secrets mode is experimental and untested; behavior and interface may change.');
+    const secretName = 'config_json_gcp';
+    const gcpSecretFile = `/tmp/secrets-gcp-${nextTempId()}.json`;
```
[the remark this answers is not in the document]
Comment (Xain3) [reply scripts/generate-compose.js:361]: Working on #32
